Add more ML tutorials (#9344)
opensearch-project/documentation-website: _ml-commons-plugin/tutorials/vector-operations/optimize-compression.md

---
layout: default
title: Optimizing vector search using Cohere compressed embeddings
parent: Vector operations
grand_parent: Tutorials
nav_order: 20
---

# Optimizing vector search using Cohere compressed embeddings

This tutorial shows you how to optimize vector search using Cohere compressed embeddings in OpenSearch 2.19 or later. These embeddings allow for more efficient storage and faster retrieval of vector representations, making them ideal for large-scale search applications.

This tutorial uses the Cohere Embed Multilingual v3 on Amazon Bedrock. For more information about using Cohere compressed embeddings on Amazon Bedrock, see [this blog post](https://aws.amazon.com/about-aws/whats-new/2024/06/amazon-bedrock-compressed-embeddings-cohere-embed/).

> In this tutorial, you'll use the following OpenSearch features:

Author: ```suggestion
In this tutorial, you'll use the following OpenSearch components:
```

  "status": "CREATED",
  "model_id": "t64OPpUBX2k07okSZc2n"
}
```

Test the model to ensure it's working correctly:

```json
POST _plugins/_ml/models/t64OPpUBX2k07okSZc2n/_predict
{
  "parameters": {
    "texts": ["Say this is a test"],
    "embedding_types": [ "int8" ]
  }
}
```
{% include copy-curl.html %}

The response contains the generated embeddings:

```json
{
  "inference_results": [
    {
      "output": [
        {
          "name": "response",
          "dataAsMap": {
            "id": "db07a08c-283d-4da5-b0c5-a9a54ef35d01",
            "texts": [
              "Say this is a test"
            ],
            "embeddings": {
              "int8": [
                [
                  -26.0,
                  31.0,
                  ...
                ]
              ]
            },
            "response_type": "embeddings_by_type"
          }
        }
      ],
      "status_code": 200
    }
  ]
}
```

## Step 2: Create an ingest pipeline

An ingest pipeline lets you process documents before indexing them. In this case, you'll use it to generate embeddings for the `title` and `description` fields in your data.

There are two ways to set up the pipeline:

> - [Invoke the model separately for `title` and `description`](#option-1-invoke-the-model-separately-for-title-and-description): This option sends separate requests for each field, generating independent embeddings.

Author: ```suggestion
1. [Invoke the model separately for `title` and `description`](#option-1-invoke-the-model-separately-for-title-and-description): This option sends separate requests for each field, generating independent embeddings.
```

> - [Invoke the model once by combining `title` and `description`](#option-2-invoke-the-model-once-by-combining-title-and-description): This option concatenates the fields into a single input and sends one request, generating a single embedding that represents both.

Author: ```suggestion
1. [Invoke the model once by combining `title` and `description`](#option-2-invoke-the-model-once-by-combining-title-and-description): This option concatenates the fields into a single input and sends one request, generating a single embedding that represents both.
```

{"title":"The Great Gatsby","author":"F. Scott Fitzgerald","description":"A novel of decadence and excess in the Jazz Age, exploring themes of wealth, love, and the American Dream.","publication_year":1925,"genre":"Classic Fiction"}
{"index":{"_index":"books_binary_embedding"}}
{"title":"To Kill a Mockingbird","author":"Harper Lee","description":"A powerful story of racial injustice and loss of innocence in the American South during the Great Depression.","publication_year":1960,"genre":"Literary Fiction"}
{"index":{"_index":"books_binary_embedding"}}
{"title":"Pride and Prejudice","author":"Jane Austen","description":"A romantic novel of manners that follows the character development of Elizabeth Bennet as she learns about the repercussions of hasty judgments and comes to appreciate the difference between superficial goodness and actual goodness.","publication_year":1813,"genre":"Romance"}
```
{% include copy-curl.html %}

### Step 5.3: Create a search pipeline

Create a new search pipeline named `ml_inference_pipeline_cohere_search_binary` using the same configuration as in [Step 2](#step-4-search-the-index), but replacing all occurrences of `int8` with `binary`.

> 1. change `embeddings.int8[0]` to `embeddings.binary[0]`

Reviewer: ```suggestion
1. Change `embeddings.int8[0]` to `embeddings.binary[0]`.
```

> 1. change `"embedding_types": ["int8"]` to `"embedding_types": ["binary"]`

Reviewer: ```suggestion
1. Change `"embedding_types": ["int8"]` to `"embedding_types": ["binary"]`.
```

        }
      }
    }
  ]
}
```
{% include copy-curl.html %}

### Rewriting the query in the search pipeline

Create another search pipeline that rewrites the query:

```json
PUT _search/pipeline/ml_inference_pipeline_cohere_search_binary2
{
  "request_processors": [
    {
      "ml_inference": {
        "model_id": "t64OPpUBX2k07okSZc2n",
        "input_map": [
          {
            "texts": "$..match.description.query"
          }
        ],
        "output_map": [
          {
            "query_vector": "embeddings.binary[0]"
          }
        ],
        "model_config": {
          "input_type": "search_query",
          "embedding_types": ["binary"]
        },
        "query_template": """
          {
            "query": {
              "knn": {
                "description_embedding": {
                  "vector": ${query_vector},
                  "k": 10
                }
              }
            },
            "_source": {
              "excludes": [
                "title_embedding",
                "description_embedding"
              ]
            },
            "size": 2
          }
        """
      }
    }
  ]
}
```
{% include copy-curl.html %}

> Then you can use the search pipeline to run vector search the same way as in [Step 4](#step-4-search-the-index).

Reviewer: ```suggestion
Then you can use the search pipeline to run a vector search, as described in [Step 4](#step-4-search-the-index).
```
